applications/FluidDynamicsApplication/custom_utilities/fluid_auxiliary_utilities.cpp

```diff
@@ -343,6 +373,113 @@ FluidAuxiliaryUtilities::ModifiedShapeFunctionsFactoryType FluidAuxiliaryUtiliti
+        max_id = block_for_each<MaxReduction<IndexType>>(r_comm.LocalMesh().MasterSlaveConstraints(), [&](MasterSlaveConstraint& rMPC){
+            return rMPC.Id();
+        });
+    }
+    max_id = r_data_comm.MaxAll(max_id);
+
+    // Create the slip nodal MPCs
+    //TODO: We should make the id increment MPI compatible
+    if (r_comm.LocalMesh().NumberOfNodes() != 0) {
+        const IndexType dim = rModelPart.GetProcessInfo()[DOMAIN_SIZE];
+
+        // Create a map with the component unknowns
+        // Note that this is done in order to access the KratosComponents once
+        std::map<IndexType, const Variable<double>*> unknowns_map;
+        unknowns_map[0] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_X");
+        unknowns_map[1] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_Y");
+        if (dim == 3) {
+            unknowns_map[2] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_Z");
+        }
+
+        // Allocate auxiliary containers
+        array_1d<double,3> unit_normal;
+        DofPointerVectorType slave_dofs;
+        DofPointerVectorType master_dofs;
+        VectorType constant_vector(1);
+        MatrixType relation_matrix(1,dim-1);
+        ModelPart::MasterSlaveConstraintContainerType mpcs_container;
+        slave_dofs.reserve(1);
+        master_dofs.reserve(dim-1);
+        mpcs_container.reserve(rModelPart.NumberOfNodes());
+
+        // Loop nodes to create the slip MPCs
+        for (auto& r_node : rModelPart.Nodes()) {
+            if (r_node.Is(rSlipFlag)) {
+                // Calculate the nodal unit normal
+                unit_normal = r_node.FastGetSolutionStepValue(NORMAL);
+                const double norm = norm_2(unit_normal);
+                if (norm > 1.0e-12) {
+                    unit_normal /= norm;
+                } else {
+                    KRATOS_WARNING("CreateSlipMultiPointConstraints") << "Node " << r_node.Id() << " norm " << norm << " is close to zero." << std::endl;
+                }
+
+                // Find the slave DOF
+                // Note that we consider as slave DOF the one corresponding to the largest normal component
+                double max_n = 0.0;
+                IndexType slave_component;
+                for (IndexType d = 0; d < dim; ++d) {
+                    if (std::abs(unit_normal[d]) > max_n) {
+                        slave_component = d;
+                        max_n = std::abs(unit_normal[d]);
+                    }
+                }
+
+                // Set the MPC data
+                IndexType aux_j = 0;
+                slave_dofs.push_back(r_node.pGetDof(*(unknowns_map[slave_component])));
+                const double wall_normal_v = rWallNormalVelocityCalculator(r_node, unit_normal);
+                for (IndexType d = 0; d < dim; ++d) {
+                    if (d != slave_component) {
+                        constant_vector(aux_j) = wall_normal_v/unit_normal(slave_component);
```

Comment:
i would say this is not correct.

slave_dof = W*master_dofs + b

so if you have only one slave dof, than b has size 1.

Reply:
You're right. It must be out of the loop. Thanks for spotting.

Reply:
Corrected.